### PR TITLE
Voxel paint

### PR DESCRIPTION
**Pull request by thoys**
# QA Test
- Test this with multiple people.
- This QA requires HMD and handcontrollers
- Use this build
1. Test Lead: Run script https://rawgit.com/thoys/hifi/voxelPaint/unpublishedScripts/marketplace/voxelPaint/createVoxelPaint.js using `Edit-> Running Scripts -> From URL`
2. Test Lead: Click or trigger the arrows to initialize the voxelPaint platform and rezzer.
3. Each user: trigger the palette shaped rezzer to receive a palette and brush.
4. Each user: test if the different brushes and colors work, and if they're selectable.
5. Each user: stress test the rezzer, try rapidly triggering the palette rezzer, each exessive palette should delete within 20 seconds.
6. Each user: try un-equipping either the palette and tools, in both orders, and notice that the palette and tool both disappears when that is done.
7. Each user: try to lose tracking for the hand that is drawing, this should not make any glitchy lines that move towards the avatars hips. After regaining tracking, a re-trigger is required to start drawing again.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 765973d72a6a58340c9fe29e8e0dbe5c6c119788 available at: https://gist.github.com/983068b104dda3a760b9d1b563473067<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10125-765973d.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10125-765973d.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: d3acf8a1183f27549aa531809684fd3438089f58 available at: https://gist.github.com/983068b104dda3a760b9d1b563473067<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10125-d3acf8a.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10125-d3acf8a.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: e91d609fe550b821ddb9d29a8d69d441c201a14f available at: https://gist.github.com/983068b104dda3a760b9d1b563473067<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10125-e91d609.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10125-e91d609.dmg

**Comment by Nex-Pro** [comment]
## Will be testing this shortly as per request.

**Comment by Nex-Pro** [comment]
Testing

**Comment by Nex-Pro** [comment]
## PASS

Had to test this twice as it proved necessary to move assets to s3. As per Thoys.

Test this with multiple people. **DONE** One Vive user one Oculus.

This QA requires HMD and handcontrollers**DONE**

Use this build**DONE**

Test Lead: Import https://rawgit.com/thoys/hifi/voxelPaint/unpublishedScripts/marketplace/voxelPaint/voxel_painter.json using Edit-> Import Entities from URL**DONE**

Test Lead: Click or trigger the box to initialize the voxelPaint platform and rezzer.**DONE**

Each user: trigger the rezzer to receive a palette and brush. **SUCCESS**

Each user: test if the different brushes and colors work, and if they're selectable. **SUCCESS**

Each user: stress test the rezzer, try rapidly triggering the palette rezzer, each exessive palette should delete within 20 seconds. **SUCCESS**

Each user: try un-equipping either the palette and tools, in both orders, and notice that the palette and tool both disappears when that is done.  **SUCCESS**

Each user: try to lose tracking for the hand that is drawing, this should not make any glitchy lines that move towards the avatars hips. After regaining tracking, a re-trigger is required to start drawing again. **SUCCESS**

File: unpublishedScripts/marketplace/voxelPaint/handyAttacher.js [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+                },
+                dynamic: 0,
+                name: 'voxel paint palette',
+                rotation: {
+                    w: 0.89465177059173584,
+                    x: 0.022446036338806152,
+                    y: 0.43398189544677734,
+                    z: -0.10347139835357666
+                },
+                script: Script.resolvePath('voxel-paint-palette.js') + '?t=' + Date.now(),
```
Comment (huffman) [inline unpublishedScripts/marketplace/voxelPaint/handyAttacher.js]: Can we remove the cache buster?

File: unpublishedScripts/marketplace/voxelPaint/voxel-paint-aether.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    this.leaveEntity = function() {
+
+        Messages.sendLocalMessage('Hifi-Hand-Drop', 'both');
+        // While leaving the aether disable the hands short
+        /*Script.setTimeout(function() {
+            Messages.sendLocalMessage('Hifi-Hand-Disabler', 'both');
+            Script.setTimeout(function() {
+                Messages.sendLocalMessage('Hifi-Hand-Disabler', 'none');
+            }, 500);
+        }, 500);*/
```
Comment (huffman) [inline unpublishedScripts/marketplace/voxelPaint/voxel-paint-aether.js]: dead code?

File: unpublishedScripts/marketplace/voxelPaint/voxel-paint-init.js [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+//
+//  Distributed under the Apache License, Version 2.0.
+//  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+//
+
+/* global Entities, Vec3, Script, paintBucketColors */
+
+(function () {
+    Script.include('/~/system/libraries/utils.js');
+    Script.include('voxel-paint-shared.js?v14');
```
Comment (huffman) [inline unpublishedScripts/marketplace/voxelPaint/voxel-paint-init.js]: Remove the v14?

File: unpublishedScripts/marketplace/voxelPaint/voxel-paint-init.js [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+        this.addPaletteSpawner(position);
+    };
+
+    this.startNearTrigger = function (entityID) {
+        this.activate();
+    };
+
+    this.stopNearTrigger = function (entityID) {
+    };
```
Comment (huffman) [inline unpublishedScripts/marketplace/voxelPaint/voxel-paint-init.js]: remove?

File: unpublishedScripts/marketplace/voxelPaint/voxel-paint-init.js [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+    this.clickDownOnEntity = function (entityID, mouseEvent) {
+        if (!mouseEvent.isLeftButton || Settings.getValue("io.highfidelity.isEditting")) {
+            return;
+        }
+        this.activate();
+    };
+
+    this.clickReleaseOnEntity = function (entityID, mouseEvent) {
+    };
```
Comment (huffman) [inline unpublishedScripts/marketplace/voxelPaint/voxel-paint-init.js]: remove?

File: unpublishedScripts/marketplace/voxelPaint/voxel-paint-palette.js [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+//  Created by Thijs Wenker on 4/3/2017
+//  Copyright 2017 High Fidelity, Inc.
+//
+//
+//  Distributed under the Apache License, Version 2.0.
+//  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+//
+
+(function() {
+    Script.include('voxel-paint-shared.js?v14');
```
Comment (huffman) [inline unpublishedScripts/marketplace/voxelPaint/voxel-paint-palette.js]: Remove v14?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 96c2d4329ef7b6afb04802eaa51bbc785ebdc364 available at: https://gist.github.com/983068b104dda3a760b9d1b563473067<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10125-96c2d43.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10125-96c2d43.dmg

**Comment by PhilipRosedale** [comment]
@thoys are we ready to re-test this one?  Is it in the marketplace?

**Comment by thoys** [comment]
Hi @PhilipRosedale , it is ready for another round of QA since there were a lot of changes between now and the last QA, I already uploaded the scripts to marketplace as concept this is the script link to get the marketplace URLs in all assets http://mpassets.highfidelity.com/8049e3c8-1cf5-4268-b608-91795d186008-v1/createVoxelPaint.js . @huffman told me he will CR this tomorrow.

**Comment by PhilipRosedale** [comment]
@thoys I tried it and it worked without errors for me, except that I often got a different color than the one I dipped the tip of the tool into.

**Comment by thoys** [comment]
@PhilipRosedale , hmm all 12 texture drips match up their paint texture for me, the ones with a texture don't set the texture but a representing color on the brush tool. The old palette had 11, (purple is new). 

Could it be that you possibly got a cached previous version?
In what kind of a motion did you move away from the palette? 

Maybe I should have it vibrate slightly whenever there is a color or tool change.

File: unpublishedScripts/marketplace/voxelPaint/handyAttacher.js [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+//  Distributed under the Apache License, Version 2.0.
+//  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+//
+
+(function() {
+    var shared = Script.require('./voxel-paint-shared.js');
+
+    // TODO: find a better way to preload entity scripts:
+    Script.include(shared.PALETTE_ENTITY_SCRIPT);
+    Script.include(shared.TOOL_ENTITY_SCRIPT);
```
Comment (thoys) [inline unpublishedScripts/marketplace/voxelPaint/handyAttacher.js]: Yeah, I was looking in to that, that seems like a good idea!

File: unpublishedScripts/marketplace/voxelPaint/handyAttacher.js [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+    // Prefetch all the textures that are being used for the Voxels, to give a smoother experience:
+    shared.PALETTE_COLORS.forEach(function(color) {
```
Comment (thoys) [inline unpublishedScripts/marketplace/voxelPaint/handyAttacher.js]: Yeah, I agree.

File: unpublishedScripts/marketplace/voxelPaint/handyAttacher.js [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+                    w: 0.89465177059173584,
+                    x: 0.022446036338806152,
+                    y: 0.43398189544677734,
+                    z: -0.10347139835357666
+                },
+                shapeType: 'none',
+                type: 'Model'
+            });
+            Script.setTimeout(function() {
+                Messages.sendLocalMessage('Hifi-Hand-Grab', JSON.stringify({hand: attachHand, entityID: attachmentEntity}));
```
Comment (thoys) [inline unpublishedScripts/marketplace/voxelPaint/handyAttacher.js]: Yes, especially the first time an entity is created, the script used to load slower. But I believe that it doesn't instantly have a running entity-script, and this is very important for it to work with the handControllerGrab script and detect when you attach it. Also it deletes both the palette and tools on detachment of either.

**Pull request closed without merging**